shared/src/types/key/ed25519.rs

```diff
@@ -105,34 +106,36 @@ pub fn verify_signature_raw(
 /// which is can then be checked by a validity predicate wasm.
 #[derive(Clone, Debug, BorshSerialize, BorshDeserialize)]
 pub struct SignedTxData {
-    /// The tx data bytes
-    pub data: Vec<u8>,
-    /// The signature is produced on the tx data concatenated with the tx code.
+    /// The original tx data bytes, if any
+    pub data: Option<Vec<u8>>,
+    /// The signature is produced on the tx data concatenated with the tx code
+    /// and the timestamp.
     pub sig: Signature,
 }
 
-impl SignedTxData {
-    /// Initialize a new signed transaction data.
-    pub fn new(
-        keypair: &Keypair,
-        data: Vec<u8>,
-        tx_code: impl AsRef<[u8]>,
-    ) -> Self {
-        let to_sign = [&data[..], tx_code.as_ref()].concat();
-        let sig = sign(keypair, &to_sign);
-        Self { data, sig }
+/// Sign a transaction using [`SignedTxData`].
+pub fn sign_tx(keypair: &Keypair, tx: Tx) -> Tx {
+    let to_sign = tx.to_bytes();
+    let sig = sign(keypair, &to_sign);
+    let signed = SignedTxData { data: tx.data, sig }
+        .try_to_vec()
+        .expect("Encoding transaction data shouldn't fail");
+    Tx {
+        code: tx.code,
+        data: Some(signed),
+        timestamp: tx.timestamp,
     }
+}
 
-    /// Verify that the transaction has been signed by the secret key
-    /// counterpart of the given public key.
-    pub fn verify(
-        &self,
-        pk: &PublicKey,
-        tx_code: impl AsRef<[u8]>,
-    ) -> Result<(), VerifySigError> {
-        let data = [&self.data, tx_code.as_ref()].concat();
-        verify_signature_raw(pk, &data, &self.sig)
-    }
+/// Verify that the transaction has been signed by the secret key
+/// counterpart of the given public key.
+pub fn verify_tx_sig(
+    pk: &PublicKey,
+    tx: &Tx,
+    sig: &Signature,
+) -> Result<(), VerifySigError> {
+    let data = tx.to_bytes();
+    verify_signature_raw(pk, &data, sig)
```

Comment:
I tried this verification, but it failed. The verification failed because the Tx had `SignedTxData`, but the signature had been made without `SignedTxData`.

Reply:
ah, yes, thank you very much catching and fixing this!